File: rtree/RRTree.hs

module RRTree (
) where

import qualified Zone as Zone
import Zone (Zone(..), Pos)

-- concat 2 lists but i don't care what order you put them in
rconcat :: [a] -> [a] -> [a]
rconcat [] ys = ys
rconcat (x:xs) ys = rconcat xs (x : ys)

extract_each :: [a] -> [(a, [a])]
extract_each [] = []
extract_each (key:elements) = (key, elements) : (map (\(k, es) -> (k, key : es)) $ extract_each elements)

max_childs_per_node = 4
max_leafs_per_node = 6

-- all leaves float on a pond (same height)

data Tree v = Node Zone [Tree v] | Branch Zone [Leaf v]
data Leaf v = Leaf Pos v

-- the algo for insert:
-- 1) navigate through each Tree level, picking the child that will expand least
-- 2) when we reach the Branch level:
-- 2a) if there's space in the branch, add the leaf and return this branch
-- 2b) if not, split the branch into two branches and return _both_
-- 3) at each level, insert this new tree into the old tree
-- 4) also, at each level, can the extra tree fit into this one? (what criteria? probably max childs per treenode)
-- 4a) if yes, fold it into the main tree
-- 4b) if no, keep ascending with both
-- requires functions:
-- a) insert into branch :: Leaf v -> Tree v -> [Tree v]
-- b) insert into node :: Leaf v -> Tree v -> [Tree v]

-- (result, extra_branch_that_needs_to_be_stuck_somewhere)
insert :: Leaf v -> Tree v -> Tree v
insert leaf tree = case result of
                        output : [] -> output
                        otherwise -> error $ "bad result length of insert: " ++ show (length result)
    where result = insert_leaf leaf tree
    
insert_leaf :: Leaf v -> Tree v -> [Tree v]
-- insert_leaf into a node:
-- 1) extract children
-- 2) pick the best child to insert into
-- 3) insert
-- 4) is there a result?
-- 4a) yes. can a result be combined into this tree?
-- 4aa) yes. do it
-- 4ab) no. don't
insert_leaf leaf (Node zone childs) = make_tree new_childs
    where 
        extracted_childs = extract_each childs
        best_child_pair = best_child_to_insert_into (leaf_pos leaf) extracted_childs

insert2 leaf (Node zone childs) = make_tree ((insert leaf best_child) : other_childs)
  where (best_child, other_childs) = best_thing
        extracted_childs = extract_each childs
insert2 leaf (Branch zone leafs)
  | length leafs > max_leafs_per_node = ((make_branch leafs1), leafs2)
  | otherwise = Branch (Zone.extend pos zone) (leaf : leafs)
  where (Leaf pos _) = leaf
        (leafs1, leafs2) = split (leaf : leafs)

-- this is a stupid split algorithm but it fits the interface
split :: [Leaf v] -> ([Leaf v], [Leaf v])
split leafs = split_n (length leafs / 2) leafs

split_n 0 leafs = ([], leafs)
split_n n (leaf:leafs) = (leaf : left, right)
    where (left, right) = split_n (n - 1) leafs

make_tree :: [Tree v] -> Tree v
make_tree childs = Node (Zone.amalgamate $ map tree_zone childs) childs

make_branch :: [Leaf v] -> Tree v
make_branch leafs = Branch (Zone.create_container $ map leaf_pos leafs) leafs

tree_zone :: Tree v -> Zone
tree_zone (Node zone _) = zone
tree_zone (Branch zone _) = zone

leaf_pos :: Leaf v -> Pos
leaf_pos (Leaf pos _) = pos

File: rtree/RRTree.java

import java.util.*;

public class RRTree<T> {
    public static class Pos {
        public final int x;
        public final int y;

        public Pos(int x, int y) {
            this.x = x;
            this.y = y;
        }

        public String toString() {
            return "(" + x + "," + y + ")";
        }
    }
    
    public static class Zone {
        public static final Zone none = new Zone(null, null);
        
        public final Pos nw;
        public final Pos se;

        public Zone(int w, int n, int e, int s) {
            this.nw = new Pos(w, n);
            this.se = new Pos(e, s);
        }

        public Zone(Pos nw, Pos se) {
            this.nw = nw;
            this.se = se;
        }

        public boolean contains(Pos pos) {
            if(this.nw == null || this.se == null) {
                return false;
            }
            
            if(pos.x < nw.x) return false;
            if(pos.x > se.x) return false;
            if(pos.y < nw.y) return false;
            if(pos.y > se.y) return false;
            return true;
        }

        public boolean overlaps(Zone zone) {
            if(this.nw == null || this.se == null) {
                return false;
            }
            
            if(zone.nw == null || zone.se == null) {
                return false;
            }
            
            if(this.nw.x > zone.se.x) return false;
            if(this.nw.y > zone.se.y) return false;
            if(this.se.x < zone.nw.x) return false;
            if(this.se.y < zone.nw.y) return false;
            return true;
        }

        public Zone extendTo(Pos pos) {
            Pos nw, se;
            
            if(this.nw == null) {
                nw = pos;
            } else {
                int x = Math.min(this.nw.x, pos.x);
                int y = Math.min(this.nw.y, pos.y);
                nw = new Pos(x, y);
            }
            
            if(this.se == null) {
                se = pos;
            } else {
                int x = Math.max(this.se.x, pos.x);
                int y = Math.max(this.se.y, pos.y);
                se = new Pos(x, y);
            }

            return new Zone(nw, se);
        }

        public Zone extendTo(Zone zone) {
            return this.extendTo(zone.nw).extendTo(zone.se);
        }

        public String toString() {
            return "[Zone " + nw + ", " + se + "]";
        }
    }

    public static abstract class Tree<T> {
        public Tree<T> parent;
        public Zone zone;
        
        public abstract void insert(Leaf<T> leaf);
        public abstract void upsert(Tree<T> node);
    }
    
    public static class Leaf<T> {
        public final Pos pos;
        public final T value;
        
        public Leaf(Pos pos, T value) {
            this.pos = pos;
            this.value = value;
        }
    }

    public static class Page<T> extends Tree<T> {
        public final ArrayList<Leaf<T>> leafs = new ArrayList<>();

        public static final int maxPageSize = 6;

        public void insert(Leaf<T> leaf) {
            leafs.add(leaf);
            if(leafs.size() > maxPageSize) {
                Page<T> newPage = split();
                if(parent == null) {
                    parent = new Node<T>();
                    parent.upsert(this);
                } else {
                    extendZone(leaf.pos);
                }
                parent.upsert(newPage);
            }
        }

        private Page<T> split() {
            Page<T> page = new Page<T>();
            for(int i = 0; i < leafs.size(); ++i) {
                Leaf<T> leaf = this.leafs.remove(leafs.size() - 1);
                page.insert(leaf);
            }
            return page;
        }

        public void upsert(Tree<T> node) {
            if(parent == null) {
                parent = new Node<T>();
                parent.upsert(this);
            }
            parent.upsert(node);
        }
    }

    public static class Node<T> extends Tree<T> {
        public final ArrayList<Tree<T>> childs = new ArrayList<>();

        public static final int maxChildsPerNode = 3;

        public void upsert(Tree<T> node) {
            if(childs.size() < maxChildsPerNode) {
                childs.add(node);
                node.parent = this;
                extendZone(node.zone);
            } else {
                if(parent == null) {
                    parent = new Node<T>();
                    parent.upsert(this);
                }
                parent.upsert(node);
            }
        }

        public void insert(Leaf<T> leaf) {
            Tree<T> child = chooseNearestChild(leaf.pos);
            child.insert(leaf);
            refreshZone();
        }

        private void refreshZone() {
            this.zone = nodesZone(this.childs);
        }

        public static Zone nodesZone(Iterable<Tree<T>> nodes) {
            Zone zone = Zone.none;
            for(Tree<T> node : nodes) {
                zone.extend(node.zone);
            }
            return zone;
        }
    }
}
